src/crimson/os/seastore/cache.cc

```diff
@@ -28,6 +28,24 @@ Cache::~Cache()
   ceph_assert(extents.empty());
 }
 
+Cache::retire_extent_ret Cache::retire_extent_if_cached(
+  Transaction &t, paddr_t addr)
+{
+  if (auto ext = t.write_set.find_offset(addr); ext != t.write_set.end()) {
+    t.add_to_retired_set(CachedExtentRef(&*ext));
```

Comment:
i am a little bit concerned with the boxing-unboxing dance. 

on the cache's consumer side, i think we always interact the cache with CachedExtents (either logical or physical). but when we clone or release an extent, we end up with a `inc_ref(txn, paddr)` call.

but to retire the extent, we have to create another `CachedExtents` instance with the value of `CachedExtent` in question. can we just capture and pass the `LogicalCachedExtentRef` being retired down to `Cache::retire_extent_if_cached()` along with its paddr ?

Reply:
This operation is a noop if the extent in question isn't already in cache -- it does not create new CachedExtent's.

That line is creating a CachedExtentRef, not a CachedExtent.  The CachedExtentRef constructor is simply a refcount increment (it's an alias for intrusive_ptr).  The reason for invocation in the first place is that write_set is intrusive and returns a bare reference to the contained value.

If you call dec_ref on an actual LogicalCachedExtentRef, it skips these lookups.  This pathway happens when you need to dec_ref an extent you don't have a reference to.  For instance, if a sequence of extents are overwritten, you don't need to actually read them, you can simply release them and move on.  The only catch is that if they happen to be cached, you need to invalidate them.  That's what this method does.

Reply:
i see.`&*iter` evaluates to a raw pointer. i misread it.